CIP-XXX/CIP-XXX.md

```diff
@@ -27,10 +27,53 @@ As part of this, we also describe the bit ordering within a `BuiltinByteString`,
 and provide some laws these operations should obey.
 
 ## Motivation: why is this CIP necessary?
+Currently, Plutus Core has a `BuiltinByteString` type, which is a packed array of bytes, similar to Haskell's ByteString.
+We want to implement bit wise operations on `BuiltinByteString` type. In otherwords, we want to achieve bit level granularity on the operations on `BuiltinByteString`.
+Some of the operations that we are interested in are, bit level OR, AND, NOT, or XOR. There are many added advantage of having bit level granularity which we will discuss in more detail later. Unfortunately, all oprations available on the `BuiltinByteString` works on the byte level granularity. So we need this CIP to bridge the gap.
+To explain the usefulness of this granulaity to manipulate `BuiltinByteString`, we use the following example of integer sets.
+#### Integer sets
+Integer sets, also known as bitmap, bitset, or bitvector is used to represent a set of natural numbers. Integer sets can be represented as a `BuiltinByteString` of length $N$, where $N\in \mathbb{N}$ is the maximum possible number in that set. We can represent membership of an element $k \leq N$ by setting the $k^{th}$ bit is set in that `BuiltinByteString`.  For example, assume $ x = \{1,3,4,5\}$ be an integer set. We can represent this set using a byte string of size $8$ as $01011100$. Note that the $1^{st}$, $3^{rd}$, $4^{th}$ and $5^{th}$ bits are set and rest are zero. There are many advantages when we represet a set using Integer set. 
+1. Representing the empty set and the universe. We can use the `0x00`  or $00000000$ to represent an empty set, and `0xFF` or $11111111$ as the universe that can be represetend using $8$ bit byte string. 
+2. We can do set intersection using bit wise AND operation. Assume two sets $x = \{2,5,6\}$ and $y = \{2,4,5\}$, then $x \cap y = \{2,5\}$. The following example shows how we can represent set union using the bit wise AND operation.
+    ```mermaid 
+    graph LR 
+    A[00100110] --> C[Bitwise AND]
+    B[00101100] --> C
+    C --> D[00100100]
+    
+    ```
+3. We can do set union by using bit wise OR operation. Let us use the same sets $x$ and $y$. Then the unition $x \cup y = \{2,4,5,6\}$. The following image shows how the bitwise OR operation can be used to compute this union.
+    ```mermaid 
+    graph LR 
+    A[00100110] --> C[Bitwise OR]
+    B[00101100] --> C
+    C --> D[00101110]
+    
+    ```
+
+4. We can do complement using bit wise NOT (assuming finite universe). Let us take the set $x = \{2,5,6\}$, and assume that the universal set be $\{0,1,2,3,4,5,6,7\}$. Then, the complement of the set $x$, is $\neg x = \{0,1,3,4,7\}$. The following figure represents shows how we can represent these sets in byte strings and how bitwise NOT operation can be used to compute the complement of a set.
+    ```mermaid 
+    graph LR 
+    A[00100110] --> C[Bitwise NOT]
+    C --> D[11011001]
+    
+    ```
```

Comment:
The universe is _guaranteed_ finite: the set can only represent values inside a fixed range. Also, once again, examples are probably not needed here, and bit order again.